edb/edgeql/compiler/stmtctx.py

```diff
@@ -204,9 +205,12 @@ def fini_expression(
             context=srcctx,
         )
 
+    used_params = {p.name for p in irutils.get_parameters(ir)}
```

Comment:
This is inefficient because it searches the entire AST tree with a callback on every node.  Let's use an accumulator in the context instead, adding every time an `ir.Parameter` is constructed.

Reply:
This is how `query_parameters` work. But transformation that makes it unused happens later in the transformation as far as I understand (i.e. unused ones are present in `query_parameters`).

Reply:
In `typegen.py:82` make a copy of `query_parameters` in `subctx` and discard it afterward. That way you will ensure that only used parameters end up in `query_parameters`

Reply:
Could there be any other operators that discard values of arguments?

Reply:
Could there be an optimization pass that discard things like `0*$x`?

Reply:
Not currently, no.